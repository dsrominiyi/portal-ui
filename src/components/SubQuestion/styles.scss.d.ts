export const subQuestionContainer: string;
export const insetQuestionWrapper: string;
export const enter: string;
export const enterActive: string;
export const exit: string;
export const exitActive: string;
