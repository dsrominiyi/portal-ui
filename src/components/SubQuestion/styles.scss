@import '../../style/vars';

$transition-ms: 200ms;

.sub-question-container {
  display: flex;
  flex-direction: column;
  height: 0;
  margin-left: 20px;
  overflow: hidden;
  transition-duration: $transition-ms;
  transition-property: height, margin, opacity;
}

.inset-question-wrapper {
  border-left: 2px solid $brand-grey-white;
  padding-left: 10px;
  transition: height $transition-ms;

  > * {
    margin: 20px 0;
  }

  > :first-child {
    margin-top: 0;
  }

  > :last-child {
    margin-bottom: 0;
  }
}

.enter {
  opacity: 0;
  margin-left: 0;
}

.enter-active {
  opacity: 1;
  margin-left: 20px;
}

.exit {
  opacity: 1;
  margin-left: 20px;
}

.exit-active {
  opacity: 0;
  margin: 0; // remove any margins applied from other sources for a smoother transition
}
