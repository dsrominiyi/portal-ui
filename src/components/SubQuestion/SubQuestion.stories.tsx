import React, { ReactNode } from 'react';
import { storiesOf } from '@storybook/react';
import SubQuestion from '.';
import AffixedField from '../TextField/AffixedField';
import QuestionBox from '../QuestionBox';

import { formSection } from '../../forms/styles.scss';

const stories = storiesOf('SubQuestion', module);

const firstInputField = <AffixedField key="0" prefix={'\u00A3'} />;
const secondInputField = <AffixedField key="1" suffix="per month" />;

const singleQuestion: ReactNode[] = [
  <QuestionBox
    key="1"
    text="How much would you want to pay for your car insurance?"
    inputFields={[firstInputField]}
  />
];

const multipleQuestions: ReactNode[] = [
  <QuestionBox
    key="1"
    text="How much would you want to pay for your car insurance?"
    inputFields={[firstInputField]}
  />,
  <QuestionBox
    key="2"
    text="How often do you want to pay this amount?"
    inputFields={[secondInputField]}
  />
];

const render = (items?: ReactNode[]) => (props?: any) => (
  <SubQuestion visible={(props && props.visible) || true} items={items} />
);

const renderWithParentQuestion = (items?: ReactNode[]) => () => {
  return (
    <div className={formSection}>
      <QuestionBox
        key="1"
        text="What do you currently pay for your car insurance?"
        inputFields={[firstInputField]}
      />
      <SubQuestion visible items={items} />
    </div>
  );
};

export const storyRenders = {
  single: render(singleQuestion),
  multiple: render(multipleQuestions),
  withParentQuestion: renderWithParentQuestion(multipleQuestions)
};

stories.add('Single Sub-question', storyRenders.single);
stories.add('Multiple Sub-questions', storyRenders.multiple);
stories.add('With parent question', storyRenders.withParentQuestion);
