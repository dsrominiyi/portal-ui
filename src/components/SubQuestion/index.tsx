import React, { FunctionComponent, ReactNode, useEffect, useRef } from 'react';
import { CSSTransition } from 'react-transition-group';

import styles from './styles.scss';

interface Props {
  visible?: boolean;
  items?: ReactNode[];
}

// transition time should match the one defined in the stylesheet
const transitionMs = 200;

export const SubQuestion: FunctionComponent<Props> = ({ visible, items }: Props) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const itemsRef = useRef<HTMLDivElement>(null);

  const setHeight = () => {
    const containerDiv = containerRef.current as HTMLInputElement;
    const itemsDiv = itemsRef.current as HTMLInputElement;
    containerDiv.style.height = visible ? `${itemsDiv.scrollHeight}px` : '';
    containerDiv.style.overflow = visible ? 'visible' : 'hidden';
  };

  const observer = new MutationObserver(setHeight);

  useEffect(() => {
    const containerDiv = containerRef.current as HTMLInputElement;
    if (containerDiv) {
      setHeight();
      observer.observe(containerDiv, { childList: true, subtree: true });
    }

    return () => observer.disconnect();
  }, [visible]);

  return (
    <CSSTransition
      in={visible}
      timeout={transitionMs}
      classNames={{
        enter: styles.enter,
        enterActive: styles.enterActive,
        exit: styles.exit,
        exitActive: styles.exitActive
      }}
      mountOnEnter
      unmountOnExit
    >
      <div ref={containerRef} className={styles.subQuestionContainer}>
        <div ref={itemsRef} className={styles.insetQuestionWrapper}>
          {items}
        </div>
      </div>
    </CSSTransition>
  );
};

SubQuestion.displayName = 'SubQuestion';
export default SubQuestion;
