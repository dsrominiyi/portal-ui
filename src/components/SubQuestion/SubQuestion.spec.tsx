import { mount, ReactWrapper } from 'enzyme';
import { storyRenders } from './SubQuestion.stories';

import testStorySnapshots from '../../helpers/test/testStorySnapshots';

let component: ReactWrapper;
const mutationObserverInstance: MutationObserver = (MutationObserver as jest.Mock)();

describe('SubQuestion', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    component = mount(storyRenders.single({ visible: false }));
  });

  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders, ['withParentQuestion']);
  });

  test('the component height is set when its made visible', () => {
    component.setProps({ visible: true });
    let containerDiv = component.find('.subQuestionContainer');

    // scroll height is 0 in jsdom
    expect(containerDiv.getDOMNode().getAttribute('style')).toBe('height: 0px; overflow: visible;');

    component.setProps({ visible: false });
    containerDiv = component.find('.subQuestionContainer');

    expect(containerDiv.getDOMNode().getAttribute('style')).toBe('overflow: hidden;');
  });

  test('the component height is set when dom changes occur', () => {
    expect(MutationObserver).toHaveBeenCalledWith(expect.any(Function));
    expect(mutationObserverInstance.observe).toHaveBeenCalledWith(expect.any(Node), {
      childList: true,
      subtree: true
    });
  });
});
