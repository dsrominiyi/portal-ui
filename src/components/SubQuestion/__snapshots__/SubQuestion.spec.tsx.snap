// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubQuestion component story renders match snapshots: multiple 1`] = `
<CSSTransition
  classNames={
    Object {
      "enter": "enter",
      "enterActive": "enterActive",
      "exit": "exit",
      "exitActive": "exitActive",
    }
  }
  in={true}
  mountOnEnter={true}
  timeout={200}
  unmountOnExit={true}
>
  <div
    className="subQuestionContainer"
  >
    <div
      className="insetQuestionWrapper"
    >
      <QuestionBox
        inputFields={
          Array [
            <AffixedField
              prefix="£"
            />,
          ]
        }
        key="1"
        text="How much would you want to pay for your car insurance?"
      />
      <QuestionBox
        inputFields={
          Array [
            <AffixedField
              suffix="per month"
            />,
          ]
        }
        key="2"
        text="How often do you want to pay this amount?"
      />
    </div>
  </div>
</CSSTransition>
`;

exports[`SubQuestion component story renders match snapshots: single 1`] = `
<CSSTransition
  classNames={
    Object {
      "enter": "enter",
      "enterActive": "enterActive",
      "exit": "exit",
      "exitActive": "exitActive",
    }
  }
  in={true}
  mountOnEnter={true}
  timeout={200}
  unmountOnExit={true}
>
  <div
    className="subQuestionContainer"
  >
    <div
      className="insetQuestionWrapper"
    >
      <QuestionBox
        inputFields={
          Array [
            <AffixedField
              prefix="£"
            />,
          ]
        }
        key="1"
        text="How much would you want to pay for your car insurance?"
      />
    </div>
  </div>
</CSSTransition>
`;
